Data Layer Refactor and Handler Tests

RageCage64 opened the pull request:
This is the first PR in the effort to settle #58 

Writing unit tests for the handlers led to a number of challenges related to testing with GORM, which prompted a refactor of the data layer. Instead of exposing the gorm DB object from the `database` package, expose a `UserStore` interface that defines the actions we will use to get User data. This separates data access concerns to the data layer, which subsequently makes the handler testing a lot more sane. 

Wrote the first set of handler tests with a mocked data layer, which tests most common cases for the handlers using `httptest.Recorder` and a test `gin.Context`. These tests can be used as blueprints for contributors adding future handlers.

RageCage64 committed:
handlers: added ping handler unit test

RageCage64 committed:
database, handlers, middleware: Store Interfaces

Refactored database package to exposing a Store interface for data
interaction instead of exposing the DB. The primary inspiration for this
was to allow sane unit testing of handlers, but also separates concerns
of data access from the handlers.

Changed the usages of database.DB in handlers and middleware to using an
exported GORM User Store.

Fixed potential panic in Me handler, check for successful cast.

RageCage64 committed:
handlers: Added user handler tests

Using a mocked UserStore interface, wrote Unit Tests for the handlers in
users.go. Also ran go mod tidy.

RageCage64 committed:
handlers: Return all user data from Me

Me endpoint will use the object from the database instead of the JWT to
send data back.

RageCage64 committed:
handlers: Appease golangci-lint

io.NopCloser appears to also work but technically it's in the ioutil
package so whatever this is fine idk what the difference is.

RageCage64 committed:
handlers: Check error of utility function

I forgot to check the error of the function that creates the request
bodies in the tests.

RageCage64 committed:
database: Genericize UserUniquenesError message

This is the message that gets sent back to the client, who doesn't need
to know why user creation failed. The data is still there in the error
object itself if we want to perform different business logic on it
later.

RageCage64 committed:
handlers: Better fail log and JSON assertion

We should get into a habit of using t.Fatalf instead of silently
failing, so that should start with this PR. Also included an assertion
for the JSON response.

RageCage64 committed:
handlers, router: Renamed PingHandler to Ping

The other handlers were renamed to not say "Handler" so reflecting that
with Ping as well.

RageCage64 committed:
database, handlers: Setting database.Users to ref

It makes more sense for this exported member to contain a reference. We
can't technically enforce that, but hopefully future stores will follow
suit.

Refactored the mockUserStore setup in the tests to not use a global
Users table, instead making the Users table a member that is reset every
test. This is to avoid tests accidentally polluting a shared data table.

RageCage64 committed:
handlers: Added t.Fatalf failure messages

Added statusCodeMistmatchMessage failure message function and use it in
all tests for more descriptive failure logs.

RageCage64 committed:
router: Removed named handlers import

Was probably left in by accident, but the named import is redundant
because we were naming it the exact same thing as the package.

RageCage64 committed:
handlers: Added PingResponse type

We will thank ourselves later if we start typing our responses. It will
also make for more sane testing, and defines our API response contract
more explicitly.

RageCage64 committed:
handlers: Added types for user handler responses

Ass with the previous commit, adding types for each handlers responses.

RageCage64 committed:
handlers: Test user handler JSON responses

Added tests for user handler JSON responses using the new response
types.

TheTedder committed:
remove useless UserResponse struct

RageCage64 committed:
handlers: Consolidated response structs

There were some duplicated response structs in the original set, so
consolidated them into different structs for the type of data a handler
will respond with. Moved the ErrorResponse into a new common.go file,
since that same error response will likely be reused in future handlers.

tendstofortytwo reviewed (commented):
The code looks great overall! Left one small nitpick.

RageCage64 committed:
database, handlers: Remove unnecessary User in err

The User struct doesn't need to be part of the error struct as we don't
do any logic on them yet, and usually whoever receives the error has a
reference to the user anyway.

FrozenSake reviewed (commented):
Some questions/comments, and a concern around returning full user structs. I'm not sure what the best way to validate passwords would be, but creating a struct of it outside the DB feels like not the tea.

RageCage64 committed:
handlers: Testing Content-Type header

Testing the Content-Type header is the expected `application/json`,
since if the response has a JSON body it should have the expected
content type. Also added helper methods for common assertion patterns.

RageCage64 committed:
database, handlers: Remove unnecessary err fields

The UserNotFoundError also had unnecessary fields ID and Email. These
are things that callers will always have when receiving the error, so
making them part of the Error struct is superfluous.

RageCage64 committed:
database: explanatory comments, error name format

Added comments that explain the relationships between `users.go` and
`gorm_user_store.go` more clearly.

Fixed staticcheck error, failed to follow convention on error names in
previous commit.

RageCage64 committed:
handlers: Removed tests

The tests are living on a new branch based on this one. This branch is
now for the purpose of the User Store changes.

RageCage64 committed:
handlers: Use renamed ErrUserNotFound

I neglected to reflect the renamed error in handlers/user.go

RageCage64 committed:
handlers: Added necessary return function

Re-added a return statement that terminates the function as intended on
success.

RageCage64 committed:
database, handlers: Removed ErrorColumn field

The ErrorColumn field is not clear cut to retrieve from a Postgres Wire
error for a unique_violation, so rather than further complicate this PR
I simplified the UserUniquenessError to become ErrUserNotUnique.

RageCage64 committed:
handlers: ErrorResponse returns an array of error

The common ErrorResponse now responds with an array of errors in the
event that multiple errors occurred during the request.

RageCage64 committed:
handlers: Added SuccessResponse with top level key

Using `data` as a top level key for successful responses.

TheTedder reviewed (approved):
Thank you for all the work you've put in on this. The separate between the user store and the db backing it is a good feature.
